src/ray/object_manager/object_manager.cc

```diff
@@ -466,45 +462,53 @@ void ObjectManager::ReceivePushRequest(std::shared_ptr<TcpClientConnection> conn
+  RAY_LOG(DEBUG) << "ReceivePushRequest " << conn->GetClientID() << " " << object_id
+                 << " " << chunk_index;
   RAY_CHECK_OK(DequeueTransfers());
 }
 
 ray::Status ObjectManager::ExecuteReceiveObject(
-    const ClientID &client_id, const ObjectID &object_id, uint64_t object_size,
+    const ClientID &client_id, const ObjectID &object_id, uint64_t data_size,
+    uint64_t metadata_size, uint64_t chunk_index,
     std::shared_ptr<TcpClientConnection> conn) {
-  boost::system::error_code ec;
-  int64_t metadata_size = 0;
-  const plasma::ObjectID plasma_id = ObjectID(object_id).to_plasma_id();
-  // Try to create shared buffer.
-  std::shared_ptr<Buffer> data;
-  std::shared_ptr<plasma::PlasmaClient> store_client = store_pool_.GetObjectStore();
-  arrow::Status s =
-      store_client->Create(plasma_id, object_size, NULL, metadata_size, &data);
-  std::vector<boost::asio::mutable_buffer> buffer;
-  if (s.ok()) {
-    // Read object into store.
-    uint8_t *mutable_data = data->mutable_data();
-    buffer.push_back(asio::buffer(mutable_data, object_size));
+  RAY_LOG(DEBUG) << "ExecuteReceiveObject " << client_id << " " << object_id << " "
+                 << chunk_index;
+
+  std::pair<const ObjectBufferPool::ChunkInfo &, ray::Status> pair = buffer_pool_.CreateChunk(object_id, data_size, metadata_size, chunk_index);
+  ObjectBufferPool::ChunkInfo chunk_info = pair.first;
+  if (pair.second.ok()) {
+    // Avoid handling this chunk if it's already being handled by another process.
+    std::vector<boost::asio::mutable_buffer> buffer;
+    buffer.push_back(asio::buffer(chunk_info.data, chunk_info.buffer_length));
+    boost::system::error_code ec;
     conn->ReadBuffer(buffer, ec);
-    if (!ec.value()) {
-      ARROW_CHECK_OK(store_client->Seal(plasma_id));
-      ARROW_CHECK_OK(store_client->Release(plasma_id));
+    if (ec.value() == 0) {
+      buffer_pool_.SealChunk(object_id, chunk_index);
     } else {
-      ARROW_CHECK_OK(store_client->Release(plasma_id));
-      ARROW_CHECK_OK(store_client->Abort(plasma_id));
-      RAY_LOG(ERROR) << "Receive Failed";
+      buffer_pool_.AbortCreateChunk(object_id, chunk_index);
+      // TODO(hme): This chunk failed, so create a pull request for this chunk.
     }
   } else {
-    RAY_LOG(ERROR) << "Buffer Create Failed: " << s.message();
+    RAY_LOG(ERROR) << "Buffer Create Failed: " << pair.second.message();
     // Read object into empty buffer.
-    std::vector<uint8_t> mutable_data;
-    mutable_data.resize(object_size + metadata_size);
-    buffer.push_back(asio::buffer(mutable_data, object_size + metadata_size));
+    uint64_t buffer_length = buffer_pool_.GetBufferLength(chunk_index, data_size);
+    std::vector<uint8_t> mutable_vec;
+    mutable_vec.resize(buffer_length);
+    uint8_t *mutable_data = mutable_vec.data();
+    std::vector<boost::asio::mutable_buffer> buffer;
+    buffer.push_back(asio::buffer(mutable_data, buffer_length));
```

Comment:
`asio::buffer` also has a [constructor](http://think-async.com/Asio/asio-1.10.6/doc/asio/reference/buffer.html) that can take in the `mutable_vec` vector directly. Also, probably should've pointed this out earlier, but is it necessary to have the `std::vector<boost::asio::mutable_buffer> buffer`? Why not pass in the mutable buffer directly?